openapi.go

```diff
@@ -227,25 +206,69 @@ func RegisterOpenAPIOperation[T, B any](s *Server, method, path string) (*openap
 	return operation, nil
 }
 
-func tagFromType(v any) string {
+type schemaTag struct {
+	openapi3.SchemaRef
+	name string
+}
+
+func schemaTagFromType[V any](s *Server, v any) (schemaTag, error) {
 	if v == nil {
-		return "unknown-interface"
+		return schemaTag{
+			name: "unknown-interface",
+			SchemaRef: openapi3.SchemaRef{
+				Ref: "#/components/schemas/unknown-interface",
+			},
+		}, nil
 	}
 
-	return dive(reflect.TypeOf(v), 4)
+	return dive[V](s, reflect.TypeOf(v), schemaTag{}, 5)
 }
 
-// dive returns the name of the type of the given reflect.Type.
-// If the type is a pointer, slice, array, map, channel, function, or unsafe pointer,
-// it will dive into the type and return the name of the type it points to.
-func dive(t reflect.Type, maxDepth int) string {
+func dive[V any](s *Server, t reflect.Type, tag schemaTag, maxDepth int) (schemaTag, error) {
+	if maxDepth == 0 {
+		return schemaTag{
+			name: "default",
+			SchemaRef: openapi3.SchemaRef{
+				Ref: "#/components/schemas/default",
+			},
+		}, nil
+	}
+
 	switch t.Kind() {
-	case reflect.Ptr, reflect.Slice, reflect.Array, reflect.Map, reflect.Chan, reflect.Func, reflect.UnsafePointer:
-		if maxDepth == 0 {
-			return "default"
+	case reflect.Ptr, reflect.Map, reflect.Chan, reflect.Func, reflect.UnsafePointer:
+		return dive[V](s, t.Elem(), tag, maxDepth-1)
+
+	case reflect.Slice, reflect.Array:
+		item, err := dive[V](s, t.Elem(), tag, maxDepth-1)
+		if err != nil {
+			return schemaTag{}, err
 		}
-		return dive(t.Elem(), maxDepth-1)
+
+		tag.name = item.name
+		tag.Value = &openapi3.Schema{
+			Type:  "array",
+			Items: &item.SchemaRef,
+		}
+		return tag, nil
+
 	default:
-		return t.Name()
+		var err error
+		tag.name = t.Name()
+		tag.Ref = "#/components/schemas/" + tag.name
+		tag.Value, err = s.getOrCreateSchema(tag.name, new(V))
+		return tag, err
+	}
+}
+
+func (s *Server) getOrCreateSchema(key string, v any) (*openapi3.Schema, error) {
+	var err error
+	schemaRef, ok := s.OpenApiSpec.Components.Schemas[key]
+	if !ok {
+		schemaRef, err = generator.NewSchemaRefForValue(v, s.OpenApiSpec.Components.Schemas)
```

Comment:
need test for this case

Reply:
It is not possible for this function error unless a custom schema customizer is provided or `ThrowErrorOnCycle` is set.